php_phongo_structs.h

```diff
@@ -284,6 +284,12 @@ typedef struct {
 	zend_object std;
 } php_phongo_commandsucceededevent_t;
 
+typedef struct {
+	bson_oid_t                topology_id;
+	const mongoc_host_list_t* host;
```

Comment:
Related to my later comment in ServerOpeningEvent.c, [`mongoc_apm_server_opening_get_host`](http://mongoc.org/libmongoc/1.17.5/mongoc_apm_server_opening_get_host.html) instructs us to copy this if we need it to live beyond the lifetime of libmongoc's APM event object (which we do).

`mongoc_apm_server_opening_get_host` also tells us that the return value is not actually a list and it's just using this type to represent a single host (i.e. the `next` pointer should be `NULL`). That means we need only worry about copying a single struct. I'll also note that the [`mongoc_host_list_t`](http://mongoc.org/libmongoc/1.17.5/mongoc_host_list_t.html) is part of libmongoc's public API, unlike some of the other opaque types we've worked with.

I think we have two options for `phongo_apm_server_opening`:

 * Declare this as a single `mongoc_host_list_t` (not a pointer) and use [`memcpy`](https://man7.org/linux/man-pages/man3/memcpy.3.html) to copy the contents of the `mongoc_apm_server_opening_get_host`'s return value (dereferencing the pointer) into this struct.
 * Declare separate fields for the host string and port based on the same declarations in `mongoc_host_list_t`'s definition, and copy those values accordingly. Since the host string has a fixed size (`BSON_HOST_NAME_MAX + 1`), you won't even need to allocate/free it.

I think either approach is fine. The choice you make will also impact the functions in ServerOpeningEvent.c that need to access the host and port, so in that sense it may be easier to use the second style and avoid needing to work with a `mongoc_host_list_t` in that file.

Reply:
Got it, I'll do the second approach then